Small iterator improvements

mrks opened the pull request:
* only create transform_iterator when really needed 
* Remove ConstantValueIterable

```
+-----------+------------------+------+------------------+------+--------+-----------------------------------------------+
| Benchmark | prev. iter/s     | runs | new iter/s       | runs | change |               p-value (significant if <0.001) |
+-----------+------------------+------+------------------+------+--------+-----------------------------------------------+
| TPC-H 1   | 0.61164188385    | 4    | 0.686662852764   | 4    | +12%   | (run time too short) (not enough runs) 0.0274 |
| TPC-H 2   | 12.1166000366    | 61   | 14.5197095871    | 73   | +20%   |                   (run time too short) 0.0000 |
| TPC-H 3   | 8.86163234711    | 45   | 7.07546806335    | 36   | -20%   |                   (run time too short) 0.0000 |
| TPC-H 4   | 0.144625231624   | 1    | 0.427503228188   | 3    | +196%  |    (run time too short) (not enough runs) nan |
| TPC-H 5   | 8.68201732635    | 44   | 8.11193943024    | 41   | -7%    |                   (run time too short) 0.0040 |
| TPC-H 6   | 25.3605232239    | 127  | 24.5789661407    | 123  | -3%    |                   (run time too short) 0.0311 |
| TPC-H 7   | 0.476894438267   | 3    | 0.485608786345   | 3    | +2%    | (run time too short) (not enough runs) 0.8362 |
| TPC-H 8   | 2.32841825485    | 12   | 1.83067798615    | 10   | -21%   |                   (run time too short) 0.0001 |
| TPC-H 9   | 1.62554419041    | 9    | 1.22846543789    | 7    | -24%   | (run time too short) (not enough runs) 0.0000 |
| TPC-H 10  | 7.30904150009    | 37   | 6.24166202545    | 32   | -15%   |                   (run time too short) 0.0000 |
| TPC-H 11  | 14.5244226456    | 73   | 17.3515739441    | 87   | +19%   |                   (run time too short) 0.0000 |
| TPC-H 12  | 5.16605806351    | 26   | 5.3621931076     | 28   | +4%    |                   (run time too short) 0.4611 |
| TPC-H 13  | 9.87368583679    | 50   | 9.08228969574    | 46   | -8%    |                   (run time too short) 0.0285 |
| TPC-H 14  | 10.0825681686    | 51   | 11.4646587372    | 58   | +14%   |                   (run time too short) 0.0000 |
| TPC-H 15  | 0.0496799573302  | 1    | 0.0538278743625  | 1    | +8%    |    (run time too short) (not enough runs) nan |
| TPC-H 16  | 2.23444080353    | 12   | 2.1240363121     | 11   | -5%    |                   (run time too short) 0.0007 |
| TPC-H 17  | 1.26653039455    | 7    | 1.34616148472    | 7    | +6%    | (run time too short) (not enough runs) 0.1233 |
| TPC-H 18  | 0.0133905988187  | 1    | 0.0134532181546  | 1    | +0%    |                         (not enough runs) nan |
| TPC-H 19  | 0.310646176338   | 2    | 0.334314733744   | 2    | +8%    | (run time too short) (not enough runs) 0.3683 |
| TPC-H 20  | 0.00459289830178 | 1    | 0.00456940662116 | 1    | -1%    |                         (not enough runs) nan |
| TPC-H 21  | 0.0716575831175  | 1    | 0.243637785316   | 2    | +240%  |    (run time too short) (not enough runs) nan |
| TPC-H 22  | 0.274164229631   | 2    | 0.297675251961   | 2    | +9%    | (run time too short) (not enough runs) 0.2298 |
| average   |                  |      |                  |      | +20%   |                                               |
+-----------+------------------+------+------------------+------+--------+-----------------------------------------------+
```

mjendruk commented on src/lib/storage/vector_compression/fixed_size_byte_aligned/fixed_size_byte_aligned_vector.hpp:
```diff
@@ -42,17 +42,31 @@ class FixedSizeByteAlignedVector : public CompressedVector<FixedSizeByteAlignedV
+    if constexpr (std::is_same_v<UnsignedIntType, uint32_t>) {
+      return _data.cend();
+    } else {
+      return boost::make_transform_iterator(_data.cend(), cast_to_uint32);
```
I had hoped that the iterator would get optimised away. Too bad.

mrks commented:
I removed the test because I removed the iterator that it tested.

`_unary_scan_with_value` is the core of all TableScans that don't use a second column, so it's already tested.

I requested your review to make sure that someone looks at this. Other than @mrzzzrm, most people only do a review when they are explicitly asked, so I am starting to do that. No particular reason why I chose you this time (and no direct criticism in your direction either).

mrks commented:
Anything else or approve plz?

mrks commented:
Ok, that particular function is not tested, but I think this is in line with the granularity of most other tests.

For funsies, I added a `!` before `func`, thus inverting all matches. This led to a failure for 332 out of 2019 tests. Guess we are safe. ;)
